horovod/common/operations.cc

```diff
@@ -1099,16 +1159,39 @@ void BackgroundThreadLoop(HorovodGlobalState& state, MPIContext& ctx) {
     auto horovod_autotune_log = std::getenv(HOROVOD_AUTOTUNE_LOG);
     state.param_manager.Initialize(rank, RANK_ZERO, ctx.mpi_comm,
                                    horovod_autotune_log != nullptr
-                                       ? std::string(horovod_autotune_log)
-                                       : "");
+                                   ? std::string(horovod_autotune_log)
+                                   : "");
     state.param_manager.SetAutoTuning(true);
   }
 
   // Initialize the tensor count table. No tensors are available yet.
   if (is_coordinator) {
-    state.message_table = std::unique_ptr<MessageTable>(new MessageTable());
+    state.message_table = std::unique_ptr<MessageTable>(
+        new MessageTable());
   }
 
+#if COMPILE_WITH_GLOO
+  auto data_layer = std::getenv("HOROVOD_CPU_OPERATIONS");
```

Comment:
What happens to hierarchical algorithms if HOROVOD_CPU_OPERATIONS=gloo is set?  Would they use Gloo + NCCL?

Reply:
They currently use MPI still, because each of those operations is hardcoded to its API calls (in other words, they're not abstracted behind an interface).

Reply:
It's typically 5 secongs longer. With gloo it takes around 95s, and without gloo it takes 90s.

Reply:
Yes. For hierarchical NCCL, we are still using MPI for inter-machine communications as hard coded.